core/src/Kokkos_Core_fwd.hpp

```diff
@@ -330,6 +332,131 @@ struct LAnd;
+template <>
+struct IsSpaceAvailable<Cuda> : std::false_type {};
+template <>
+struct IsSpaceAvailable<CudaSpace> : std::false_type {};
+template <>
+struct IsSpaceAvailable<CudaHostPinnedSpace> : std::false_type {};
+template <>
+struct IsSpaceAvailable<CudaUVMSpace> : std::false_type {};
+#endif
+
+namespace Experimental {
+class HBWSpace;
+}  // namespace Experimental
+#if !defined(KOKKOS_ENABLE_HBWSPACE)
+template <>
+struct IsSpaceAvailable<Experimental::HBWSpace> : std::false_type {};
+#endif
+
+namespace Experimental {
+class OpenMPTarget;
+class OpenMPTargetSpace;
+}  // namespace Experimental
+#if !defined(KOKKOS_ENABLE_OPENMPTARGET)
+template <>
+struct IsSpaceAvailable<Experimental::OpenMPTarget> : std::false_type {};
+template <>
+struct IsSpaceAvailable<Experimental::OpenMPTargetSpace> : std::false_type {};
+#endif
+
+namespace Experimental {
+class HIP;
+class HIPSpace;
+class HIPHostPinnedSpace;
+}  // namespace Experimental
+#if !defined(KOKKOS_ENABLE_HIP)
+template <>
+struct IsSpaceAvailable<Experimental::HIP> : std::false_type {};
+template <>
+struct IsSpaceAvailable<Experimental::HIPSpace> : std::false_type {};
+template <>
+struct IsSpaceAvailable<Experimental::HIPHostPinnedSpace> : std::false_type {};
+#endif
+
+namespace Experimental {
+class SYCL;
+class SYCLDeviceUSMSpace;
+class SYCLSharedUSMSpace;
+}  // namespace Experimental
+#if !defined(KOKKOS_ENABLE_SYCL)
+template <>
+struct IsSpaceAvailable<Experimental::SYCL> : std::false_type {};
+template <>
+struct IsSpaceAvailable<Experimental::SYCLDeviceUSMSpace> : std::false_type {};
+template <>
+struct IsSpaceAvailable<Experimental::SYCLSharedUSMSpace> : std::false_type {};
+#endif
+
+namespace Impl {
+template <typename... T>
+struct GatherActiveSpaces;
+
+template <class, class>
+struct Concatenate;
+template <class... First, class... Second>
+struct Concatenate<type_list<First...>, type_list<Second...>> {
+  using type = type_list<First..., Second...>;
+};
+
+template <typename... T>
+struct GatherActiveSpaces;
+template <>
+struct GatherActiveSpaces<> {
+  using type = type_list<>;
+};
+template <typename T, typename... Ts>
+struct GatherActiveSpaces<T, Ts...> {
+  using other_types = typename GatherActiveSpaces<Ts...>::type;
+  using type        = typename std::conditional<
+      IsSpaceAvailable<T>{},
+      typename Concatenate<type_list<T>, other_types>::type, other_types>::type;
+};
+}  // namespace Impl
+
+using ActiveExecutionSpaces =
+    Impl::GatherActiveSpaces<Serial, Threads, OpenMP, Cuda,
+                             Experimental::OpenMPTarget, Experimental::HIP,
+                             Experimental::SYCL>::type;
+using ActiveMemorySpaces = Impl::GatherActiveSpaces<
+    HostSpace, Experimental::HBWSpace, CudaSpace, CudaHostPinnedSpace,
+    CudaUVMSpace, Experimental::OpenMPTargetSpace, Experimental::HIPSpace,
+    Experimental::HIPHostPinnedSpace, Experimental::SYCLDeviceUSMSpace,
+    Experimental::SYCLSharedUSMSpace>::type;
```

Comment:
What are those two things used for?

Reply:
Several people were asking for an iterable list of active execution and memory spaces and that is what these two variables do. We can add that in a follow-up pull request, though.